perl_genrule.bzl

```
@@ -0,0 +1,128 @@
+
+    commands_joined, srcs_as_files, outs_as_files = generate_commands(binary_invocation, ctx.attr.assembly_flavor, ctx.attr.srcs_to_outs, ctx.attr.srcs_to_outs_dupes, ctx)
+    outs_as_files_paths = [out.path for out in outs_as_files]
+    srcs_as_files_paths = [src.path for src in srcs_as_files]
+    perl_generate_file = ctx.file._perl_generate_file
+    if ctx.attr.is_unix:
+        ctx.actions.run(
+            inputs = srcs_as_files + additional_srcs + [ctx.toolchains["@rules_perl//perl:toolchain_type"].perl_runtime.interpreter],
+            outputs = outs_as_files,
+            executable = perl_generate_file,
+            arguments = [commands_joined],
+            mnemonic = "GenerateAssemblyFromPerlScripts",
+            progress_message = "Generating files {} from scripts {}".format(outs_as_files_paths, srcs_as_files_paths),
+        )
+    else:
+        ctx.actions.run_shell(
+            inputs = srcs_as_files + additional_srcs,
+            outputs = outs_as_files,
+            command = commands_joined,
+            mnemonic = "GenerateAssemblyFromPerlScriptsOnWindows",
+            progress_message = "Generating files {} from scripts {} on Windows".format(outs_as_files_paths, srcs_as_files_paths),
+            use_default_shell_env = True,
+        )
+
+    cc_info = CcInfo(
+        compilation_context = cc_common.create_compilation_context(direct_private_headers = outs_as_files),
+    )
+    ret = [DefaultInfo(files = depset(outs_as_files)), cc_info]
+    return ret
+
+perl_genrule = rule(
+    implementation = _perl_genrule_impl,
+    doc = "Generate files using perl.",
+    attrs = {
+        # Additional sources needed by the generation scripts.
+        "additional_srcs": attr.label_list(allow_files = True, doc = "List of other input files used by the main input files."),
+        # We specify what assemby flavor to use based on os and architecture.
+        "assembly_flavor": attr.string(doc = "What flavor to use for assembly generation."),
+        # We need to know what os this is running on.
+        "is_unix": attr.bool(doc = "Whether this is mac or linux or not."),
+        # The dict of srcs to their outs.
+        "srcs_to_outs": attr.label_keyed_string_dict(allow_files = True, doc = "Dict of input to output files from their source script."),
+        # The dicts of srcs to their outs when they are dupes from the first dict.
+        "srcs_to_outs_dupes": attr.label_keyed_string_dict(allow_files = True, doc = "Dict of input to output files where the source is dupe from the first dict."),
+        # Script that handles the file generation and existence check. Only used on nix.
+        "_perl_generate_file": attr.label(
+            allow_single_file = True,
+            executable = True,
+            cfg = "exec",
+            default = "@openssl-generated-overlay//:perl_generate_file.sh",
+        ),
+    },
+    toolchains = [config_common.toolchain_type("@rules_perl//perl:toolchain_type", mandatory = False)],
```

I actually think you'll wanna consume `current_perl_toolchain` so you get an interpreter for the exec configuration.

```suggestion
        "_perl_toolchain": attr.label(
            cfg = "exec",
            default = Label("@rules_perl//perl:current_perl_toolchain"),
        ),
    },
```

---

Done! Thanks!